pkg/threescale/backend/backend.go

```diff
@@ -566,21 +566,78 @@ func (a *Application) pruneUnlimitedCounter(snapshot Application) {
+	elapsedPeriods := a.LocalState.getElapsedPeriods(a.timestamp, remoteTimestamp)
+	// find any metrics and hence associated limits that have been deleted from the updated view of the state
+	added, removed := computeAddedAndRemovedMetrics(a.LocalState, a.RemoteState)
+
+	// sync state by pruning added and removed rate limiting periods from known metrics
+	a.LocalState = synchronizeStates(a.LocalState, a.RemoteState)
+
+	// if we found a new metric in the remote state, we must add it to known local state
+	for _, newMetric := range added {
+		a.LocalState[newMetric] = a.RemoteState[newMetric]
+		if flushingState.reportingErr {
+			// if we failed to report we should check if we previously had any counters for this metric tht
+			// were unlimited and if so update the counters with the current known value
+			if currentValue, wasKnownUnlimited := a.UnlimitedCounter[newMetric]; wasKnownUnlimited {
+				for _, counters := range a.LocalState[newMetric] {
+					counters.CurrentValue += currentValue
+				}
+			}
+		}
 
-	//TODO - check for elapses and handle accordingly
-	time.Unix(flushingState.snapshot.timestamp, 0).Add(time.Minute).After(time.Unix(remoteTimestamp, 0))
+	}
 
-	a.RemoteState = remoteState
+	if !flushingState.reportingErr {
+		// reset the counter if we have reported already
+		a.pruneUnlimitedCounter(flushingState.snapshot)
+	}
 
 	for metric, counters := range a.LocalState {
 
-		if canAdjust := a.statesAreComparable(metric, 0, a.RemoteState, flushingState.snapshot.LocalState); !canAdjust {
+		// if the limit has been removed we need to convert the current values to an unlimited metric for reporting
+		// if we failed to report, add current values, if we reported successfully add the activity since
+		// remove the metric from our local state and continue
+		if contains(metric, removed) {
+			now := a.LocalState[metric][0].CurrentValue
+			if flushingState.reportingErr {
+				a.UnlimitedCounter[metric] = now
+			} else {
+				then := flushingState.snapshot.LocalState[metric][0].CurrentValue
+				diff := now - then
+				a.UnlimitedCounter[metric] = diff
+			}
+			delete(a.LocalState, metric)
 			continue
 		}
 
+		// we are now, from this point on working with a fully synchronised state
+		// all new metrics have been added to local state
+		// all delete metrics have been removed from local state
+		// all previously known metrics with additional limits have had their limits added and set as remote state
+		// all previously known metrics with removed limits have had their limits removed remote state
+		// *we can now guarantee the state are comparable once again*
+		// update the local state to reflect the updated timestamp for the next iteration
+		a.timestamp = remoteTimestamp
 		lowestLocalGranularity := counters[0]
 		lowestSnappedGranularity := flushingState.snapshot.LocalState[metric][0]
 		lowestRemoteGranularity := a.RemoteState[metric][0]
+		if reflect.DeepEqual(lowestLocalGranularity, lowestRemoteGranularity) {
```

Comment:
thats a good point it would. I guess I'm going to need to extract a further list of added and removed limits from each metrics and skip over periods that have been added, wdyt?

Reply:
yes, absolutely :D